### PR TITLE
277 switch between double and float

### PR DESCRIPTION
**Pull request by ZhenxiZhao**
switch precision TODO: simbody and eigen

### CODE REVIEW

**Comment by ZhenxiZhao** [comment]
> @ZhenxiZhao I have went through the modifications you have made in this pull request, and they are seems OK. Would you please update me of the existing issues of this improvement?

All the constants in the project (not variable, for instance, 1.0) should be rewite as Real(1.0) or ::One(). One related error is shown below,
```
/home/SPHinXsys/SPHINXsys/src/shared/materials/riemann_solver.cpp:30:77: error: no matching function for call to ‘SMAX(SPH::Real, double)’
   30 |   return rho0c0_geo_ave_ * u_jump * SMIN(3.0 * SMAX(u_jump * inv_c_ave_, 0.0), 1.0);
```

There are many cast type conversion in the project, for instance, 
```
	template <class DampingAlgorithmType>
	bool DampingWithRandomChoice<DampingAlgorithmType>::RandomChoice()
	{
		return ((double)rand() / (RAND_MAX)) < random_ratio_ ? true : false;
	}
```

Besides, the words "SimTk::xxx" should be replaced by "SimTKxxx", for instance, on line 43, solid_body_supplementary.cpp
```
			body_part_volume * solid_body_density_, SimTK::Vec3(0), SimTK::UnitInertia(Ix, Iy, Iz));

```

**Comment by Xiangyu-Hu** [comment]
> > @ZhenxiZhao I have went through the modifications you have made in this pull request, and they are seems OK. Would you please update me of the existing issues of this improvement?
> 
> All the constants in the project (not variable, for instance, 1.0) should be rewite as Real(1.0) or ::One(). One related error is shown below,
> 
> ```
> /home/SPHinXsys/SPHINXsys/src/shared/materials/riemann_solver.cpp:30:77: error: no matching function for call to ‘SMAX(SPH::Real, double)’
>    30 |   return rho0c0_geo_ave_ * u_jump * SMIN(3.0 * SMAX(u_jump * inv_c_ave_, 0.0), 1.0);
> ```
> 
> There are many cast type conversion in the project, for instance,
> 
> ```
> 	template <class DampingAlgorithmType>
> 	bool DampingWithRandomChoice<DampingAlgorithmType>::RandomChoice()
> 	{
> 		return ((double)rand() / (RAND_MAX)) < random_ratio_ ? true : false;
> 	}
> ```
> 
> Besides, the words "SimTk::xxx" should be replaced by "SimTKxxx", for instance, on line 43, solid_body_supplementary.cpp
> 
> ```
> 			body_part_volume * solid_body_density_, SimTK::Vec3(0), SimTK::UnitInertia(Ix, Iy, Iz));
> ```

Yes. Seems that we need do it.

**Comment by Xiangyu-Hu** [comment]
Actually, using replace function in VS code, such replacement is not hard.

**Comment by Xiangyu-Hu** [comment]
@ZhenxiZhao could you pull from the main repository to solve the conflicts?

**Comment by ChiZhangatTUM** [comment]
@ZhenxiZhao @Xiangyu-Hu I have moved the modification of branch "[ZhenxiZhao:277-switch-between-double-and-float]" to branch"[Xiangyu-Hu:277-switch-between-double-and-float]" , which we can continue this implementation.

**Comment by ChiZhangatTUM** [comment]
@Xiangyu-Hu Do I have the access to push to this branch "[Xiangyu-Hu:277-switch-between-double-and-float]" ?

**Comment by ChiZhangatTUM** [comment]
@ZhenxiZhao 
I encounter this error after setting "SimTk::xxx" replaced by "SimTKxxx"
Screenshot 2023-05-29 at 16 52 47

According to Simbody documentation

Screenshot 2023-05-29 at 16 57 15


Does this mean that we should build Simbody with float as the default precision or add another transfer form float vector to double vector there ?

**Comment by ZhenxiZhao** [comment]
> @ZhenxiZhao I encounter this error after setting "SimTk::xxx" replaced by "SimTKxxx" Screenshot 2023-05-29 at 16 52 47
> 
> According to Simbody documentation
> 
> Screenshot 2023-05-29 at 16 57 15
> Does this mean that we should build Simbody with float as the default precision or add another transfer form float vector to double vector there ?

Yes, I attenmpt to use the fVec or Vec explicitly, since the type is pre-defined before compling simbody in vcpkg. 
However, the class or struct in simbody use Vec, which has been pre-defined as double. Thus we encounter this error. 
I think if we can dual compile simbody to simply implement the type switch, one for float, one for double.

**Comment by ChiZhangatTUM** [comment]
> > @ZhenxiZhao I encounter this error after setting "SimTk::xxx" replaced by "SimTKxxx" Screenshot 2023-05-29 at 16 52 47
> > According to Simbody documentation
> > Screenshot 2023-05-29 at 16 57 15
> > Does this mean that we should build Simbody with float as the default precision or add another transfer form float vector to double vector there ?
> 
> Yes, I attenmpt to use the fVec or Vec explicitly, since the type is pre-defined before compling simbody in vcpkg. However, the class or struct in simbody use Vec, which has been pre-defined as double. Thus we encounter this error. I think if we can dual compile simbody to simply implement the type switch, one for float, one for double.

That would introduce extra efforts for the installation. 
I am thinking that we can use double precision in SIMBODY, while transfer single precision of Eigen vector to double vector of SIMBODY if needed.

File: CMakeLists.txt [hunk 1]
```diff
@@ -78,6 +79,9 @@ if(TARGET Boost::program_options)
     target_compile_definitions(sphinxsys_core INTERFACE BOOST_AVAILABLE)
     target_link_libraries(sphinxsys_core INTERFACE Boost::program_options)
 endif()
+if(SPHINXSYS_SINGLE)
+    add_definitions(-D ENABLE_SINGLE)
```
Comment (FabienPean-Virtonomy) [inline CMakeLists.txt:83]: `target_compile_definitions(sphinxsys_core INTERFACE SPHINXSYS_USE_FLOAT)`

File: CMakeLists.txt [hunk 2]
```diff
@@ -10,6 +10,7 @@ option(SPHINXSYS_2D "Build sphinxsys_2d library" ON)
 option(SPHINXSYS_3D "Build sphinxsys_3d library" ON)
 option(SPHINXSYS_BUILD_TESTS "Build tests" ON)
 option(SPHINXSYS_DEVELOPER_MODE "Developer mode has more flags active for code quality" ON)
+option(SPHINXSYS_SINGLE "Build using single precision or not" OFF)
```
Comment (FabienPean-Virtonomy) [inline CMakeLists.txt:13]: SPHINXSYS_USE_FLOAT

**Comment by ChiZhangatTUM** [comment]
@Xiangyu-Hu @ZhenxiZhao 
Implementation done, while the following tests fails when using float due to the large error. 

10:test_2d_dambreak_python
13:test_2d_elastic_gate
16:test_2d_eulerian_taylor_green
30:test_2d_self_contact
31:test_2d_shell
43:test_2d_oscillating_beam_UL
46:test_2d_stfb
47:test_2d_stlw
53:test_vector_functions_particle_relaxation
58:bernoulli_beam_struct_sim
76:test_3d_passive_cantilever
79:test_3d_roof
80:roof_under_self_weight.dp_4
83:test_3d_roof_parametric_cvt
90:test_3d_twisting_column

**Comment by Xiangyu-Hu** [comment]
this branch stop to use as another branch on the same function has been merged.

**Pull request closed without merging**